### ISSUE TITLE
local attention with unidirectional lstm not converging

### ISSUE THREAD

**Issue by manish-kumar-garg** [rwth-i6/returnn-experiments#41]
Hi @albertz ,
I tried with different learning rates but model seems to be not converging after changing bidirectional lstm to unidirectional lstm in case of local attention setup.
Can you suggest something?
What else should I try.

**Comment by albertz** [comment]
It seems a very related question was also asked in #42.
I have not tried unidir LSTMs in the encoder yet, so I don't know. You probably should play around with all the available hyper params, e.g.:

* Learning rate (initial at warmup, and highest learning rate after warmup)
* Learning rate warmup length (num epochs)
* Pretraining. E.g. start number of layers (try 2). Initial time reduction (try increasing it, e.g. 6, 8, 16, or even 32). Try make it longer (more repetitions). Etc.
* No or less SpecAugment in the beginning.
* Higher batch size in the beginning. Or gradient accumulation in the beginning.
* Curriculum learning, i.e. the `epoch_wise_filter` option.
* ...

Let this smallest network with highest time reduction, high batch size, less/no SpecAugment etc train like that for as long as needed, before increasing anything. This small network should first get some half-way good score. Only when you see that, at that point the pretraining can increase the depth and other things, but only carefully and slowly (such that the network does not totally break again).

**Comment by manish-kumar-garg** [comment]
Seems like decreasing the learning rate helps.
Also, increasing the lstm cell size to 1536 helps a bit, however, not much.

What other combination do you suggest to try next?

**Comment by albertz** [comment]
All the things which I wrote already ([here](https://github.com/rwth-i6/returnn-experiments/issues/41#issuecomment-574067091)), but basically everything else as well.

**Comment by manish-kumar-garg** [comment]
Lowering the learning rate  `lr=0.0005` and `lr_init=0.0002` worked for me.
Thanks!
